docs/specs/tsd-011-internal-plugin-statistics.md

```diff
@@ -0,0 +1,51 @@
+# Internal plugin statistics collection
+
+## Objective
+
+Provide a way to report plugin-internal statistics for collection by the
+[`internal` input plugin][internal].
+
+[internal]: /plugins/inputs/internal/README.md
+
+## Keywords
+
+plugins, statistics
+
+## Overview
+
+The [`internal` input plugin][internal] allows to collect statistics about
+active plugins in Telegraf. These statistics are valuable indicators for
+operating and optimizing Telegraf setups and for detecting issues.
+
+Statistics are provided by plugin models and are then gathered by the
+[`internal` input plugin][internal] to be emitted through the normal plugin
+pipeline. However, not all important statistics are known on the model level.
+Some are only known within the plugin instance itself such as the bytes written
+to an output or certain error types. Emitting those statistics through direct
+registration of a statistics object will not pick-up tags defined using the
+global tags settings or aliases which are only known at the model level.
+
+To overcome the mentioned limitation this specification defines a framework
+allowing to inject a _statistics collector_ object into the plugin allowing to
+register statistics including the `alias` and tags definitions.
+
+To provide plugin-internal statistics a plugin should export a `Statistics`
+member in the plugin structure. This member must be a pointer type
+`*selfstat.Collector`. It is strongly recommended to define a `"-"` TOML tag
+in order to avoid collisions with setting the member through user configuration.
+
+The model code then must inject a `selfstat.Collector` instance into the
```

Comment:
Can we be more specific about the injection timing?
"The model code must inject a `selfstat.Collector` instance into the plugin's `Statistics` member after plugin instantiation but before calling the `Init` function."